### ISSUE TITLE
Support for Lian-Li Uni Fan SL-INFINITY fan hub

### ISSUE THREAD

**Issue by CUexter** [liquidctl/liquidctl#629]
### Device type

Fan hub/controller

### Product page

https://lian-li.com/product/uni-fan-sl-infinity/

### First-party software

L-Connect 3

### What monitoring functionality does the device support?

temperatures, fan/pump speeds or duty cycles

### What configuration functionality does the device support?

fan/pump speeds or duty cycles, lighting of accessories like RGB fans or LED strips

### Connection protocol

USB

### Additional information about the device

[It seems like this can be used to help ?](https://github.com/EightB1ts/uni-sync)

**Comment by LuNeder** [comment]
Not sure if I should open a new issue for this, but please do the V2 too! (the linked program seems to also support it so I assume the method should be quite similar?)

**Comment by ThimbleFire** [comment]
This would be great. The only software I'm familiar with that can run these operates only on Windows, and is Rem0o's FanControl. My fans will likely be stuck at 650 RPM until this gets added since they can't be controlled by the bios.

**Comment by LuNeder** [comment]
> This would be great. The only software I'm familiar with that can run these operates only on Windows, and is Rem0o's FanControl. My fans will likely be stuck at 650 RPM until this gets added since they can't be controlled by the bios.

@ThimbleFire the [software op mentioned in their issue](https://github.com/EightB1ts/uni-sync) seem to have worked for me to set my fans to 100%, had to try it after my cpu was overheating when no software was controlling the fans and it seems to have done the job bc that didn’t happen anymore (and it’s a little bit more noisy too)

Tho yeah liquidctl support would be really really great since then we’d have better control and the ability to use stuff like coolero

**Comment by Serphentas** [comment]
The utility is licensed with MIT so it should not pose any problem with respect to code usage. I will work on an initial draft but I need someone who can test it, any volunteers ?

**Comment by LuNeder** [comment]
> The utility is licensed with MIT so it should not pose any problem with respect to code usage. I will work on an initial draft but I need someone who can test it, any volunteers ?

I could definitely test it!

**Comment by Serphentas** [comment]
OK I've got the initial port from Rust done, see https://github.com/Serphentas/liquidctl/commit/3d018aba76e8af81616564b5ae52d276a0a7b791. I have to check a few things with Jonas and then I will ask you to kindly test it. Thank you for your patience.

**Comment by Serphentas** [comment]
I think the code is ready to be tested now, you may pull [this branch](https://github.com/Serphentas/liquidctl/tree/feat-lianli-uni). Let me know how things go.

**Comment by LuNeder** [comment]
> I think the code is ready to be tested now, you may pull [this branch](https://github.com/Serphentas/liquidctl/tree/feat-lianli-uni). Let me know how things go.

I'll try it this weekend and let you know, thanks!

**Comment by LuNeder** [comment]
Hey, got to test it!

The device seems to be listed correctly
````
python3 -m liquidctl list                     
Device #0: LianLi-UNI SL
````
but it errors when trying to get the status:
````
sudo python3 -m liquidctl status        
ERROR: LianLi-UNI SL: unexpected error: NotImplementedError()
````

**Comment by Serphentas** [comment]
Thanks for your feedback. Forgot to mention only the `set fanX speed N` function is supported right now. The source repo I used did not showcase any kind of status probing, so that was left out.

Can you then test `liquidctl set fan1 speed 100` ? You may add `--sync-rgb` to sync RGB (or not, leave it out) and `--fan-mode {rpm,pwm}` where `pwm` will not use the provided speed percentage.

**Comment by LuNeder** [comment]
Oh, oke!!

Now I get:
````
sudo python3 -m liquidctl set fan1 speed 100
ERROR: LianLi-UNI SL: unexpected error: ValueError('fan channel must be between 1 and 6')
````

I tried other stuff instead of "fan1" (such as numbers between 1 and 6), but the same error happens

**Comment by Serphentas** [comment]
Yeah I've made a change to that part, I had misunderstood the expected value for channel. Would you mind pulling and trying again ?

**Comment by LuNeder** [comment]
New error now:

````
sudo python3 -m liquidctl set fan1 speed 100
ERROR: LianLi-UNI SL: unexpected error: KeyError('fan_mode')
````

**Comment by CUexter** [comment]
any further update ? for this issue ?

**Comment by Luminatrix-eng** [comment]
Hi, my apologies for reviving a closed conversation. I was wondering if anyone here can help me with an issue. I am hoping to control my Lian Li uni fans through CoolerControl which depends on liquidctl. Currently, it looks like the liquidctl hardware support list doesn't show Lian Li as a supported device but I saw this thread with support for Lian-Li Uni. And typing 'liquidctl list' in the terminal doesn't show the Lian Li fans. 

How do I add the code mentioned above? And is there any further update for this issue? I'm a bit of a noob when it comes to this.
